cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,667 @@
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// Default to adding a precise match if possible. Otherwise
+	// let seccomp figure out the architecture specifics.
+	if err = secFilter.AddRuleConditionalExact(secSyscall, seccomp.ActAllow, conds); err != nil {
+		err = secFilter.AddRuleConditional(secSyscall, seccomp.ActAllow, conds)
+	}
+
+	return err
+}
+
+// used to mock in tests
+var (
+	archUbuntuArchitecture       = arch.UbuntuArchitecture
+	archUbuntuKernelArchitecture = arch.UbuntuKernelArchitecture
+)
+
+// For architectures that support a compat architecture, when the
+// kernel and userspace match, add the compat arch, otherwise add
+// the kernel arch to support the kernel's arch (eg, 64bit kernels with
+// 32bit userspace).
+func addSecondaryArches(secFilter *seccomp.ScmpFilter) error {
+	// note that all architecture strings are in the dpkg
+	// architecture notation
+	var compatArch seccomp.ScmpArch
+
+	// common case: kernel and userspace have the same arch. We
+	// add a compat architecture for some architectures that
+	// support it, e.g. on amd64 kernel and userland, we add
+	// compat i386 syscalls.
+	if archUbuntuArchitecture() == archUbuntuKernelArchitecture() {
+		switch archUbuntuArchitecture() {
+		case "amd64":
+			compatArch = seccomp.ArchX86
+		case "arm64":
+			compatArch = seccomp.ArchARM
+		case "ppc64":
+			compatArch = seccomp.ArchPPC
+		}
+	} else {
+		// less common case: kernel and userspace have different archs
+		// so add a compat architecture that matches the kernel. E.g.
+		// an amd64 kernel with i386 userland needs the amd64 secondary
+		// arch added to support specialized snaps that might
+		// conditionally call 64bit code when the kernel supports it.
+		// Note that in this case snapd requests i386 (or arch 'all')
+		// snaps. While unusual from a traditional Linux distribution
+		// perspective, certain classes of embedded devices are known
+		// to use this configuration.
+		compatArch = UbuntuArchToScmpArch(archUbuntuKernelArchitecture())
+	}
+
+	if compatArch != seccomp.ArchInvalid {
+		return secFilter.AddArch(compatArch)
+	}
+
+	return nil
+}
+
+func hasLine(content []byte, needle string) bool {
+	for _, line := range bytes.Split(content, []byte("\n")) {
+		if bytes.Equal(bytes.TrimSpace(line), []byte(needle)) {
+			return true
+		}
+	}
+	return false
+}
+
+func compile(content []byte, out string) error {
+	var err error
+	var secFilter *seccomp.ScmpFilter
+
+	// FIXME: right now complain mode is the equivalent to unrestricted.
+	// We'll want to change this once we seccomp logging is in order.
+	if hasLine(content, "@unrestricted") || hasLine(content, "@complain") {
```

Comment:
Since the seccomp profile is just an unordered list of syscalls, @unrestricted and @complain are treated similarly and may appear elsewhere. By convention in snapd we put them at the top of the file, but anywhere in the file is fine and consistent with the C implementation. We could change this of course if it is deemed a problem. I like the property because it allows for these or future directives to potentially be composed (eg, an interface that adds `@something`).